flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/source/enumerator/AbstractIcebergEnumerator.java

```diff
@@ -93,6 +98,28 @@ public void handleSourceEvent(int subtaskId, SourceEvent sourceEvent) {
     }
   }
 
+  @Override
+  public void handleSourceEvent(int subTaskId, int attemptNumber, SourceEvent sourceEvent) {
```

Comment:
Currently, the way to ensure the same split assignment for concurrent execution attempts is following:

1. Because all the splits assignment are through `SplitEnumeratorContext.assignSplits()` call, the `SourceCoordinator` actually tracks the splits assigned to a subtask, and it broadcast all the splits assignment to all the attempts.
2. On the split request side, regardless of which attempt of a subtask sends the split request, the request will be sent to the `SplitEnumerator`. That means the fastest attempt may result in new split assignment to slower attempts before they have finished consumption of the already assigned splits.

Because of the the above logic, for sources that do not have additional custom source events, they simply need to implement the `SupportsHandleExecutionAttemptSourceEvent` interface with the no-op behavior in the `handleSourceEvent` method.

In our case, the only custom `SourceEvent` is `org.apache.iceberg.flink.source.split.SplitRequestEvent`. Although the splits assignment is done via this custom `SourceEvent`, it seems that the same approach can still apply. More specifically, there is no need to distinguish between the `SplitRequestEvent` from different concurrent attempts, because a `SplitRequestEvent` basically means the fastest attempts has finished the assigned splits and ask for more. As long as the newly assigned splits are broadcasted to all the concurrent attempts, the assignment will be consistent across all the attempts. 

In sight of that, I think we can just delegate the impl of this method to `handleSourceEvent()` without worrying about the attempt id.

Reply:
Ah I see. Interesting. Got it. The exact same implementation that you described is what I saw with DynamicFileSplitEnumerator (https://github.com/apache/flink/blob/master/flink-connectors%2Fflink-connector-files%2Fsrc%2Fmain%2Fjava%2Forg%2Fapache%2Fflink%2Fconnector%2Ffile%2Fsrc%2Fimpl%2FDynamicFileSplitEnumerator.java). Makes sense now.

Reply:
Thanks for pointing it out!